### PR TITLE
New api features belkot

### PR DESCRIPTION
**Pull request by siriniok**

### CODE REVIEW

File: lib/kittikrb/cursor.rb [hunk 1]
```diff
@@ -103,6 +103,37 @@ def move_to(x, y)
...
+
+      def up(y = 1)
+        @y += y.floor
+
+        self
+      end
+
+      def down(y = 1)
+        @y -= y.floor
+
```
Comment (siriniok) [inline lib/kittikrb/cursor.rb]: Same here.

File: lib/kittikrb/cursor.rb [hunk 2]
```diff
@@ -103,6 +103,37 @@ def move_to(x, y)
...
 
+      def move_by(x, y)
+        @x += x.floor
+        @y += y.floor
+
+        self
+      end
+
+      def up(y = 1)
+        @y += y.floor
```
Comment (siriniok) [inline lib/kittikrb/cursor.rb]: Maybe, it looks like moving cursor up, but actually, you are moving it down.

File: lib/kittikrb/cursor.rb [hunk 3]
```diff
@@ -103,6 +103,37 @@ def move_to(x, y)
         self
       end
 
+      def move_by(x, y)
+        @x += x.floor
+        @y += y.floor
```
Comment (siriniok) [inline lib/kittikrb/cursor.rb]: You should detect moving direction and call relative method (left, up, etc...).

File: spec/kittikrb/cursor_spec.rb [hunk 4]
```diff
@@ -122,15 +122,219 @@ def current_rendered_buffer
...
+      describe '#move_to' do
+        it 'should properly set absolute position of cursor' do
+          expect(cursor.move_to(5, 10)).to be_instance_of described_class
+          expect(current_coords).to eq [5, 10]
+        end
+      end
+
+      describe '#move_by' do
+        it 'should properly set offset position of cursor' do
+          cursor.move_to(1, 2)
```
Comment (siriniok) [inline spec/kittikrb/cursor_spec.rb]: For which reasons?

Comment (siriniok) [inline spec/kittikrb/cursor_spec.rb]: Also, we are expecting one call of left, right, up or down method.

File: spec/kittikrb/cursor_spec.rb [hunk 5]
```diff
@@ -122,15 +122,219 @@ def current_rendered_buffer
...
+          expect(cursor.move_by(3, 7)).to be_instance_of described_class
+          expect(current_coords).to eq [4, 9]
+        end
```
Comment (siriniok) [inline spec/kittikrb/cursor_spec.rb]: We need test for moving by negative offset.

File: spec/kittikrb/cursor_spec.rb [hunk 6]
```diff
@@ -122,15 +122,219 @@ def current_rendered_buffer
...
+          expect(current_coords).to eq [-1, 0]
+        end
+      end
+
+      describe '#right' do
+        it 'should properly set right offset position of cursor' do
+          expect(cursor.right).to be_instance_of described_class
+          expect(current_coords).to eq [1, 0]
+        end
+      end
```
Comment (siriniok) [inline spec/kittikrb/cursor_spec.rb]: Previous 4 tests are very similar and contain too much of repeating code. Please, refactor them. Also, we should be sure that methods working properly with passed arguments. So we need two tests per method.

File: spec/kittikrb/cursor_spec.rb [hunk 7]
```diff
@@ -122,15 +122,219 @@ def current_rendered_buffer
...
+          cursor.move_to(1, 2)
+          expect(cursor.move_by(3, 7)).to be_instance_of described_class
+          expect(current_coords).to eq [4, 9]
+        end
+      end
+
+      describe '#up' do
+        it 'should properly set up offset position of cursor' do
+          expect(cursor.up).to be_instance_of described_class
+          expect(current_coords).to eq [0, 1]
```
Comment (siriniok) [inline spec/kittikrb/cursor_spec.rb]: `expect(current_coords).to eq [0, -1]`, isn't?

File: spec/kittikrb/cursor_spec.rb [hunk 8]
```diff
@@ -122,15 +122,219 @@ def current_rendered_buffer
...
+      describe '#move_by' do
+        it 'should properly set offset position of cursor' do
+          cursor.move_to(1, 2)
+          expect(cursor.move_by(3, 7)).to be_instance_of described_class
+          expect(current_coords).to eq [4, 9]
+        end
+      end
+
+      describe '#up' do
+        it 'should properly set up offset position of cursor' do
```
Comment (siriniok) [inline spec/kittikrb/cursor_spec.rb]: `it should properly move cursor up with default arguments`

We are writing tests for people and real world, not for machines.

File: lib/kittikrb/cursor.rb [hunk 9]
```diff
@@ -115,6 +146,113 @@ def reset_tty!
         self
       end
 
+      def erase(x1, y1, x2, y2)
```
Comment (siriniok) [inline lib/kittikrb/cursor.rb]: We need comment above this method, because there is not obvious argument list (actually, not obvious method name, but we can't change it for API compatibility).

```
   # Erase the specified region.
   # The region describes the rectangle shape which need to erase.
```

File: lib/kittikrb/cursor.rb [hunk 10]
```diff
@@ -115,6 +146,113 @@ def reset_tty!
...
+        y2.floor.downto(y1.floor) do |y|
```
Comment (siriniok) [inline lib/kittikrb/cursor.rb]: Why not `upto`?

File: lib/kittikrb/cursor.rb [hunk 11]
```diff
@@ -115,6 +146,113 @@ def reset_tty!
...
+          x1.floor.upto(x2.floor) do |x|
+            move_to x, y
```
Comment (siriniok) [inline lib/kittikrb/cursor.rb]: No, we can't move cursor.

File: lib/kittikrb/cursor.rb [hunk 12]
```diff
@@ -115,6 +146,113 @@ def reset_tty!
...
+            write_control "\s"
```
Comment (siriniok) [inline lib/kittikrb/cursor.rb]: No, we should only write one wrapped white space to the buffer, not to the tty.

File: lib/kittikrb/cursor.rb [hunk 13]
```diff
@@ -115,6 +146,113 @@ def reset_tty!
...
+      end
+
+      def eraseToStart
+        erase 0, @y, @x, @y
+
+        self
+      end
+
+      def eraseToDown
+        erase @x, 0, @x, @y
```
Comment (siriniok) [inline lib/kittikrb/cursor.rb]: Not correct.

File: lib/kittikrb/cursor.rb [hunk 14]
```diff
@@ -115,6 +146,113 @@ def reset_tty!
...
+
+        self
+      end
+
+      def eraseToUp
+        erase @x, @y, @x, @height - 1
```
Comment (siriniok) [inline lib/kittikrb/cursor.rb]: Not correct.

File: lib/kittikrb/cursor.rb [hunk 15]
```diff
@@ -115,6 +146,113 @@ def reset_tty!
...
+          x1.floor.upto(x2.floor) do |x|
+            move_to x, y
+            write_control "\s"
+          end
+        end
+
+        self
+      end
+
+      def eraseToEnd
```
Comment (siriniok) [inline lib/kittikrb/cursor.rb]: I know, it is my fault, sorry. But in ruby, all methods should be named in snake_case, not in camelCase. Please, fix it everywhere.

File: spec/kittikrb/cursor_spec.rb [hunk 16]
```diff
@@ -122,15 +122,219 @@ def current_rendered_buffer
...
+        end
+      end
+    end
+
+    context 'erase methods' do
+      let(:width) { $stdout.winsize[1] }
+      let(:height) { $stdout.winsize[0] }
+
+      describe '#erase' do
+        it 'should erase(write whitespace) the specified region in TTY' do
```
Comment (siriniok) [inline spec/kittikrb/cursor_spec.rb]: `should erase(write whitespace) the specified region in TTY`

`'should properly erase the specified region'`

Look https://github.com/kittikrb/cursor/pull/4/files#r55125092

File: spec/kittikrb/cursor_spec.rb [hunk 17]
```diff
@@ -122,15 +122,219 @@ def current_rendered_buffer
...
+      describe '#eraseScreen' do
+        it 'should erase(write whitespace) all chars in TTY' do
+          expect($stdout).to receive(:write).exactly(width * height).times
+            .with("\e\s")
+          expect(cursor.eraseScreen).to be_instance_of described_class
+        end
+      end
+    end
+
+    context 'cursor methods' do
```
Comment (siriniok) [inline spec/kittikrb/cursor_spec.rb]: Every method in this class is the cursor method :). Name this section like  'styling methods'.

File: spec/kittikrb/cursor_spec.rb [hunk 18]
```diff
@@ -122,15 +122,219 @@ def current_rendered_buffer
...
+      describe '#foreground' do
+        it 'should properly set foreground color' do
+          expect(cursor.foreground(8)).to be_instance_of described_class
```
Comment (siriniok) [inline spec/kittikrb/cursor_spec.rb]: Use symbols, not magic constants. Like `:red`.

File: spec/kittikrb/cursor_spec.rb [hunk 19]
```diff
@@ -122,15 +122,219 @@ def current_rendered_buffer
...
+        end
+      end
+
+      describe '#background' do
+        it 'should properly set background color' do
+          expect(cursor.background(5)).to be_instance_of described_class
+          expect(cursor.instance_variable_get(:@background)).to eq 5
+        end
+      end
+
```
Comment (siriniok) [inline spec/kittikrb/cursor_spec.rb]: Previous 2 tests are very similar. Please, refactor.

File: spec/kittikrb/cursor_spec.rb [hunk 20]
```diff
@@ -122,15 +122,219 @@ def current_rendered_buffer
...
+      describe '#hidden' do
+        it 'should properly set display hidden parameter' do
+          expect(cursor.hidden).to be_instance_of described_class
+          expect(cursor.instance_variable_get(:@display)[:hidden]).to be_truthy
+
+          cursor.hidden false
+          expect(cursor.instance_variable_get(:@display)[:hidden]).to be_falsey
+        end
+      end
```
Comment (siriniok) [inline spec/kittikrb/cursor_spec.rb]: Please, refactor tests for display modes.

File: spec/kittikrb/cursor_spec.rb [hunk 21]
```diff
@@ -122,15 +122,219 @@ def current_rendered_buffer
...
+
+      describe '#background' do
+        it 'should properly set background color' do
+          expect(cursor.background(5)).to be_instance_of described_class
+          expect(cursor.instance_variable_get(:@background)).to eq 5
+        end
+      end
+
+      describe '#bold' do
+        it 'should properly set display bold parameter' do
```
Comment (siriniok) [inline spec/kittikrb/cursor_spec.rb]: `should properly toggle bold mode`

https://github.com/kittikrb/cursor/pull/4/files#r55125092

File: spec/kittikrb/cursor_spec.rb [hunk 22]
```diff
@@ -122,15 +122,241 @@ def current_rendered_buffer
...
+          expect(cursor).to receive(:right).with(7)
+          expect(cursor).to receive(:down).with(3)
+          expect(cursor.move_by(7, 3)).to be_instance_of described_class
+        end
+
+        it 'should properly set negative offset position of cursor' do
+          expect(cursor).to receive(:left).with(2)
+          expect(cursor).to receive(:up).with(4)
+          expect(cursor.move_by(-2, -4)).to be_instance_of described_class
+        end
```
Comment (siriniok) [inline spec/kittikrb/cursor_spec.rb]: But there is no tests for changing cursor position at all. You should check `current_coords` after calling `move_by`.

File: spec/kittikrb/cursor_spec.rb [hunk 23]
```diff
@@ -122,15 +122,241 @@ def current_rendered_buffer
...
+
+      describe '#move_to' do
+        it 'should properly set absolute position of cursor' do
+          expect(cursor.move_to(5, 10)).to be_instance_of described_class
+          expect(current_coords).to eq [5, 10]
+        end
+      end
+
+      describe '#move_by' do
+        it 'should properly set positive offset position of cursor' do
```
Comment (siriniok) [inline spec/kittikrb/cursor_spec.rb]: We are writing tests for people and real world, not for machines. 'should properly move cursor right and down'

File: lib/kittikrb/cursor.rb [hunk 24]
```diff
@@ -115,6 +146,117 @@ def reset_tty!
...
+      end
+
+      def erase_to_start
+        erase 0, @y, @x, @y
+
+        self
+      end
+
+      def erase_to_up
+        erase @x, 0, @x, @y
```
Comment (siriniok) [inline lib/kittikrb/cursor.rb]: `0, 0, @width - 1, @y`

File: lib/kittikrb/cursor.rb [hunk 25]
```diff
@@ -115,6 +146,117 @@ def reset_tty!
...
+
+        self
+      end
+
+      def erase_to_down
+        erase @x, @y, @x, @height - 1
```
Comment (siriniok) [inline lib/kittikrb/cursor.rb]: `0, @y, @width - 1, @height - 1`

File: spec/kittikrb/cursor_spec.rb [hunk 26]
```diff
@@ -122,15 +122,82 @@ def current_rendered_buffer
...
+          expect(cursor.move_to(5, 10)).to be_instance_of described_class
+          expect(current_coords).to eq [5, 10]
+        end
+      end
+
+      describe '#move_by' do
+        it 'should properly move cursor right and down' do
+          expect(cursor).to receive(:right).with(7)
+          expect(cursor).to receive(:down).with(3)
+          expect(cursor.move_by(7, 3)).to be_instance_of described_class
```
Comment (siriniok) [inline spec/kittikrb/cursor_spec.rb:145]: Ok, I understood your problem here. We should check side effect of our stubbed methods `right` and `down`. Simply, we can just write two tests. We already have one of them. So you should write another one which will check `current_cords`. 

``` ruby
        it 'should change current coordinates' do
          cursor.move_by(7, 3)
          expect(current_coords).to eq [7, 3]
          cursor.move_by(-1, -1)
          expect(current_coords).to eq [6, 2]
        end
```

I'm sure, it is not better solution, but let's refactor it later.

File: spec/kittikrb/cursor_spec.rb [hunk 27]
```diff
@@ -122,15 +122,82 @@ def current_rendered_buffer
...
+        it 'should properly set absolute position of cursor' do
+          expect(cursor.move_to(5, 10)).to be_instance_of described_class
+          expect(current_coords).to eq [5, 10]
+        end
+      end
+
+      describe '#move_by' do
+        it 'should properly move cursor right and down' do
+          expect(cursor).to receive(:right).with(7)
+          expect(cursor).to receive(:down).with(3)
```
Comment (siriniok) [inline spec/kittikrb/cursor_spec.rb:144]: We should specify how many times we are expecting this messages.

``` ruby
expect(cursor).to receive(:right).with(7).once
expect(cursor).to receive(:down).with(3).once
```

File: lib/kittikrb/cursor.rb [hunk 28]
```diff
@@ -146,6 +146,57 @@ def reset_tty!
         self
       end
 
+      # Erase the specified region.
+      # The region describes the rectangle shape which need to erase.
+      def erase(x1, y1, x2, y2)
+        move_to x1, y1
+        length = x2.floor - x1.floor + 1
```
Comment (siriniok) [inline lib/kittikrb/cursor.rb]: First of all, it looks like very cool optimization :+1: But does it really so? Let's explore it together.

I should to say, that we can't move our cursor in this method at all. Simply, we don't have any needs to do this, really. Method `write` moves cursor, so we can't use it. Possibly, it is a problem from architectural view, and we should implement method like `write` that will operate with buffer without moving cursor, but it is not easy question (though it seems so).

Now we have only such thing like `@buffer`. And we can directly operate with it.
We can define our shape as `2, 3, 6, 7`. Or `2, 3, -6, -7`. In the last case we should normalize coordinates first than we start to erase our shape. But it is a corner case. So we can omit it now and fix it later. But. We should write our code with thinking about this problem. So we should implement this method as simple as we can.

Maybe, you heard it before, but 'premature optimization is the root of all evil' (Donald Khuth). Yes, your idea is great. But not now. It makes this code harder to rewrite in the future.

So now you should focus on simplicity of your code. On practice it means that you should use such structure:

```
        y1.floor.upto(y2.floor) do |y|
          x1.floor.upto(x2.floor) do |x|

          end
        end
```

Ok. And how we should write to the buffer? It is very easy too.

```
@buffer[get_pointer_from_xy(x, y)] = self.class.wrap ' ', { x: x, y: y }
```

These three lines will be enough for now :)

File: spec/kittikrb/cursor_spec.rb (hunk 27 above)
Comment (Belkot) [inline spec/kittikrb/cursor_spec.rb:144]: If you don't specify an expected receive count, it defaults to once.
